src/mapper/pkg/dnscache/dns_cache.go

```diff
@@ -1,46 +1,214 @@
 package dnscache
 
 import (
+	"container/list"
 	"context"
-	"github.com/jellydator/ttlcache/v3"
 	"github.com/otterize/network-mapper/src/mapper/pkg/config"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/viper"
+	"net"
+	"sync"
 	"time"
 )
 
 type DNSCache struct {
-	cache *ttlcache.Cache[string, string]
+	cache *TTLCache[string, string]
+}
+
+type Resolver interface {
+	LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error)
 }
 
 func NewDNSCache() *DNSCache {
 	capacity := viper.GetInt(config.DNSCacheItemsMaxCapacityKey)
-	dnsRecordCache := ttlcache.New[string, string](ttlcache.WithCapacity[string, string](uint64(capacity)))
-	go dnsRecordCache.Start()
-
-	lastCapacityReachedErrorPrint := time.Time{}
-	dnsRecordCache.OnEviction(func(ctx context.Context, reason ttlcache.EvictionReason, item *ttlcache.Item[string, string]) {
-		if reason == ttlcache.EvictionReasonCapacityReached && time.Since(lastCapacityReachedErrorPrint) > time.Minute {
-			logrus.Warningf("DNS cache capacity reached entries are being dropped, consider increasing config '%s'",
-				config.DNSCacheItemsMaxCapacityKey)
-			lastCapacityReachedErrorPrint = time.Now()
-		}
-	})
+	if capacity == 0 {
+		logrus.Panic("Capacity cannot be 0")
+	}
+	dnsRecordCache := NewTTLCache[string, string](capacity)
 
 	return &DNSCache{
 		cache: dnsRecordCache,
 	}
 }
 
-func (d *DNSCache) AddOrUpdateDNSData(dnsName string, ip string, ttlSeconds int) {
-	ttl := time.Duration(ttlSeconds) * time.Second
-	d.cache.Set(dnsName, ip, ttl)
+func (d *DNSCache) AddOrUpdateDNSData(dnsName string, ip string, ttl time.Duration) {
+	d.cache.Insert(dnsName, ip, ttl)
 }
 
-func (d *DNSCache) GetResolvedIP(dnsName string) (string, bool) {
+func (d *DNSCache) GetResolvedIPs(dnsName string) []string {
 	entry := d.cache.Get(dnsName)
-	if entry == nil {
-		return "", false
+	return entry
+}
+
+// CacheValue holds the value and its expiration time
+type CacheValue[V any] struct {
+	Value      V
+	Expiration time.Time
+}
```

Comment:
move the TTLCache implementation (starting here and until the last line of this file) to a separate file called `ttl_cache`. Maybe in a separate package.